File: android/src/main/kotlin/com/appodeal/appodeal_flutter/AppodealMrec.kt
package com.appodeal.appodeal_flutter

import com.appodeal.ads.MrecCallbacks
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal class AppodealMrec(
    private val flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
) : MethodChannel.MethodCallHandler {

    val adChannel: MethodChannel by lazy {
        MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "appodeal_flutter/mrec"
        ).apply { setMethodCallHandler(this@AppodealMrec) }
    }

    val adListener: Listener = Listener(adChannel)

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {}

    internal class Listener(private val adChannel: MethodChannel) : MrecCallbacks {

        override fun onMrecLoaded(isPrecache: Boolean) {
            adChannel.invokeMethod("onMrecLoaded", mapOf("isPrecache" to isPrecache))
        }

        override fun onMrecFailedToLoad() {
            adChannel.invokeMethod("onMrecFailedToLoad", null)
        }

        override fun onMrecShown() {
            adChannel.invokeMethod("onMrecShown", null)
        }

        override fun onMrecShowFailed() {
            adChannel.invokeMethod("onMrecShowFailed", null)
        }

        override fun onMrecClicked() {
            adChannel.invokeMethod("onMrecClicked", null)
        }

        override fun onMrecExpired() {
            adChannel.invokeMethod("onMrecExpired", null)
        }
    }
}

File: ios/Classes/AppodealMrec.swift
import Flutter
import Foundation
import Appodeal

internal final class AppodealMrec {
    
    internal let adChannel: FlutterMethodChannel
    internal let adListener: Listener
    
    init(registrar: FlutterPluginRegistrar) {
        adChannel = FlutterMethodChannel(name: "appodeal_flutter/mrec", binaryMessenger: registrar.messenger())
        adListener = Listener(adChannel: adChannel)
    }
    
    internal final class Listener: NSObject, APDBannerViewDelegate {
        
        private let adChannel: FlutterMethodChannel
        
        fileprivate init(adChannel: FlutterMethodChannel) {
            self.adChannel = adChannel
        }
        
        func bannerViewDidLoadAd(_ bannerView: APDBannerView, isPrecache precache: Bool) {
            if (bannerView.adSize == kAPDAdSize300x250) {
                adChannel.invokeMethod("onMrecLoaded", arguments: ["isPrecache": precache])
            }
        }
        
        func bannerViewDidShow(_ bannerView: APDBannerView) {
            if (bannerView.adSize == kAPDAdSize300x250) {
                adChannel.invokeMethod("onMrecShown", arguments: nil)
            }
        }
        
        func bannerViewExpired(_ bannerView: APDBannerView) {
            if (bannerView.adSize == kAPDAdSize300x250) {
                adChannel.invokeMethod("onMrecExpired", arguments: nil)
            }
        }
        
        func bannerViewDidInteract(_ bannerView: APDBannerView) {
            if (bannerView.adSize == kAPDAdSize300x250) {
                adChannel.invokeMethod("onMrecClicked", arguments: nil)
            }
        }
        
        func bannerView(_ bannerView: APDBannerView, didFailToLoadAdWithError error: Error) {
            if (bannerView.adSize == kAPDAdSize300x250) {
                adChannel.invokeMethod("onMrecFailedToLoad", arguments: nil)
            }
        }
    }
}
